docs: Add Dynamic tables for System Props + Compose Docs [WEB-1414] (#1641)
Codecademy/client-modules: packages/styleguide/stories/Foundations/System/Compose.stories.mdx

  }}
/>

When using custom properties or our standard <LinkTo id="foundations-system-props--page">system props</LinkTo> it can be useful to combine them into a single function using `variance.compose`.

Composing functions has several benefits:

1. Predictable CSS - Guard against possible bugs and unexpected behavior.
1. Type Simplification - Less typescript boiler plate and better type inference

## Usage

> Using `variance.compose()` will take any number of system prop functions and merge them left to right (the last argument takes precedence if props conflict) and returns a new prop function with the intersection of these pieces.

Reviewer: Breaking this up into two sentences. 1) Input, 2) Output
```suggestion
Using `variance.compose()` will take any number of system prop functions and merge them left to right (the last argument takes precedence if props conflict). `variance.compose()` returns a new prop function with the intersection of these pieces.
```

```tsx
import { variance } from '@codecademy/variance';
import { layout, grid } from '@codecademy/gamut-styles';

const Grid = styled.div(variance.compose(layout, grid));

<Grid width="100%" gridTemplateColumns="1fr 1fr" />;
```

## Why compose?

There are a few benefits to composition:

### Predictable CSS

> When Emotion serializes styles it does not attempt to merge styles and allows all styles to exist and be progressively overriden. This can create issues with styles that might conflict between scopes.

Reviewer: ```suggestion
When Emotion serializes styles, it does not attempt to merge styles. Instead, it allows all styles to exist and be progressively overridden. This can create issues with styles that might conflict between scopes.
```

> With uncombined props this can produce multiple sets of media queries that are not ordered, or in cases where 2 style functions template the same property unmerged styles.

Reviewer: ```suggestion
With uncombined props, this can produce multiple sets of media queries that are not ordered, or in cases where 2 style functions template the same property, unmerged styles.
```

```tsx
// non composed style functions
<Grid width={['100%', '80%']} justifyContent={{ xs: 'center' }} />
```

This outputs several media queries:

```css
.Grid {
  width: '100%';
}

@media (min-width: 480px) {
  .Grid {
    width: 80%;
  }
}

@media (min-width: 480px) {
  .Grid {
    justify-content: center;
  }
}
```

> However if we've composed these functions our scopes will be combined preventing unexpected behavior.

Reviewer: ```suggestion
However, if we've composed these functions, our scopes will be combined and prevent unexpected behavior.
```

  }}
/>

There may be instances where you would like to combine properties for ease of sharing and/or performance to batch their computations. To make this easy, we've added a compose function to combine returned style functions into a single one that has the intersection of all of their props.

We've provided some basic groups of props that are pre-composed through gamut-styles, which you can find <LinkTo id="foundations-system-props--page">here</LinkTo>.

> However there are many cases where we want to combine or add new props to a single function to increase performance and guard against possible property collisions. To help with this `variance` comes with a `compose` method to help with the combination.

Reviewer 2: I think you can streamline this intro a bit to reduce duplicated information:

1. Delete the first sentence
2. Move any information from that first sentence that isn't already in the last once to after the last one

Thoughts?

Author: Good call lots of duplication.  Took another pass.

## Usage

Using `variance.compose()` will take any number of system prop functions and merge them left to right (the last argument takes precedence if props conflict) and returns a new prop function with the intersection of these pieces.

```tsx
import { variance } from '@codecademy/variance';
import { layout, grid } from '@codecademy/gamut-styles';

const Grid = styled.div(variance.compose(layout, grid));

<Grid width="100%" gridTemplateColumns="1fr 1fr" />;
```

## Why compose?

There are a few benefits to composition:

### Type simplification

Composing style functions handles the type inference for you making your styled components easier to write.

If you pass more than a single function to `styled` emotion is unable to infer the intersection of both functions on its own forcing you to pass the type manually.

```tsx
// plain inference
const Grid = styled.div(layout, grid);

> <Grid width="100%" justifyContent="center" />; // typeerror justifyContent is not a prop

Reviewer 2: ```suggestion
<Grid width="100%" justifyContent="center" />; // TypeError: justifyContent is not a prop
```

// manual inference
type GridProps = StyleProps<typeof grid> & StyleProps<typeof layout>;
const Grid = styled.div<GridProps>(layout, grid);

<Grid width="100%" justifyContent="center" />; // works but we did it manually :(

// composition inference
const Grid = styled.div(variance.compose(layout, grid));

> <Grid width="100%" justifyContent="center" />; // works but no manual config :)

Reviewer 2: ```suggestion
<Grid width="100%" justifyContent="center" />; // works with no manual config :)
```

```

### Style merging

> When emotion serializes styles it does not attempt to merge styles and allows all styles to exist and be progressively overriden. This can create issues with styles that might conflict between scopes.

Reviewer 2: ```suggestion
When Emotion serializes styles it does not attempt to merge styles and allows all styles to exist and be progressively overriden. This can create issues with styles that might conflict between scopes.
```

With uncombined props this can produce multiple sets of media queries that are not ordered, or in cases where 2 style functions template the same property unmerged styles.

```tsx
// non composed style functions
<Grid width={['100%', '80%']} justifyContent={{ xs: 'center' }} />
```

> This outputs several media queries

Reviewer 2: ```suggestion
This outputs several media queries:
```

  }}
/>

When using custom properties or our standard <LinkTo id="foundations-system-props--page">system props</LinkTo> it can be useful to combine them into a single function using `variance.compose`.

Composing functions has several benefits:

1. Predictable CSS - Guard against possible bugs and unexpected behavior.
1. Type Simplification - Less typescript boiler plate and better type inference

## Usage

Using `variance.compose()` will take any number of system prop functions and merge them left to right (the last argument takes precedence if props conflict) and returns a new prop function with the intersection of these pieces.

```tsx
import { variance } from '@codecademy/variance';
import { layout, grid } from '@codecademy/gamut-styles';

const Grid = styled.div(variance.compose(layout, grid));

<Grid width="100%" gridTemplateColumns="1fr 1fr" />;
```

## Why compose?

> There are a few benefits to composition:

Reviewer 3: This organization seems a little odd to be since the colon denotes what immediately follows as benefits. so the reader might just see Predictable CSS. 

Maybe: "There are few benefits to composition, such as having [Predictable CSS](#predictable-css) and [Type Simplification](#type-simplification)" <- not sure if this linking actually works lol
